plugin-server/src/main/ingestion-queues/session-recording-v2/consumer.ts

```diff
@@ -44,39 +49,60 @@ export class SessionRecordingIngester {
     private readonly metrics: SessionRecordingMetrics
     private readonly promiseScheduler: PromiseScheduler
     private readonly batchConsumerFactory: BatchConsumerFactory
+    private readonly sessionBatchManager: SessionBatchManager
 
     constructor(
         private config: PluginsServerConfig,
         private consumeOverflow: boolean,
         batchConsumerFactory: BatchConsumerFactory,
         ingestionWarningProducer?: KafkaProducerWrapper
     ) {
+        this.topic = consumeOverflow
+            ? KAFKA_SESSION_RECORDING_SNAPSHOT_ITEM_OVERFLOW
+            : KAFKA_SESSION_RECORDING_SNAPSHOT_ITEM_EVENTS
+        this.batchConsumerFactory = batchConsumerFactory
+
         this.isDebugLoggingEnabled = buildIntegerMatcher(config.SESSION_RECORDING_DEBUG_PARTITION, true)
+
+        this.promiseScheduler = new PromiseScheduler()
+
         const kafkaMetrics = KafkaMetrics.getInstance()
         const kafkaParser = new KafkaParser(kafkaMetrics)
         const teamService = new TeamService()
         this.metrics = SessionRecordingMetrics.getInstance()
-        this.promiseScheduler = new PromiseScheduler()
-        this.batchConsumerFactory = batchConsumerFactory
+
+        const offsetManager = new KafkaOffsetManager(async (offsets) => {
+            await new Promise<void>((resolve, reject) => {
+                try {
+                    this.batchConsumer!.consumer.commitSync(offsets)
+                    resolve()
+                } catch (error) {
+                    reject(error)
+                }
+            })
+        }, this.topic)
+        this.sessionBatchManager = new SessionBatchManager({
+            maxBatchSizeBytes: (config.SESSION_RECORDING_MAX_BATCH_SIZE_KB ?? 0) * 1024,
+            maxBatchAgeMs: config.SESSION_RECORDING_MAX_BATCH_AGE_MS ?? 1000,
+            createBatch: () => new BaseSessionBatchRecorder(new BlackholeFlusher()),
+            offsetManager,
+        })
```

Comment:
This feels really complicated... I could just be missing the concept but it feels like "clever" code that is really hard to follow. I think it might be good for us to chat through it and explain your thoughts as I can't follow this at all 😅